lxd/events/devlxdEvents.go

```diff
@@ -33,31 +31,23 @@ func NewDevLXDServer(debug bool, verbose bool) *DevLXDServer {
 	return server
 }
 
-// AddListener creates and returns a new event listener.
+// AddListener creates and returns a new event listener which uses websockets.
 func (s *DevLXDServer) AddListener(instanceID int, connection *websocket.Conn, messageTypes []string) (*DevLXDListener, error) {
```

Comment:
@monstermunchkin do you think we can narrow the scope of the new interface and instead of having different implementations of the actual event listener logic, instead encapsulate the connection specific behaviors using an interface passed in to this function instead of `connection *websocket.Conn`. So instead accept a `connection EventListenerConnection` argument which exposes a common interface used by all event listener connection types.

Reply:
Something like the `EventListener` you defined, but like this:

```go
type EventListenerConnection interface {
	Reader(recvFunc EventHandler) // This would be run by the event listener server in the background, passing it the function to call if an event is received from the client. This is where the connection specific heartbeat/closed detection would be done.
	WriteJSON(ctx context.Context, event any) error // Context would be used for setting `SetWriteDeadline` internally.
	Close() // Closes the connection.
}
```

Reply:
I cannot find a nice way of doing this given the context. Basically, the `EventListenerConnection` would need to cancel the (parent) context which doesn't feel right.

We could however simplify the `EventListener` interface by removing `SetWriteDeadline`, and just set the write deadline in `WriteJSON`.

Reply:
> I cannot find a nice way of doing this given the context. Basically, the `EventListenerConnection` would need to cancel the (parent) context which doesn't feel right.

Why would the `EventListenerConnection` need to cancel the context passed to `EventListenerConnection.WriteJSON()`? It would only be used to cancel the write itself. However perhaps accepting a write deadine timeout duration on `WriteJSON()` would be easier here?

Reply:
Well as my comment said `// Context would be used for setting `SetWriteDeadline` internally.` so it *could* have taken a context. But passing a write deadline directly works too.

For the `Reader()` function, I propose that as its run in the background from the EventListener that that go routine wait until `Reader()` ends and then calls its own `Close()` function, so that the two are not tightly coupled.

Reply:
> Well as my comment said // Context would be used for setting SetWriteDeadline internally. so it could have taken a context.

I don't understand. Why does `SetWriteDeadline` need a context? It takes a `time.Time`.

> For the Reader() function, I propose that as its run in the background from the EventListener that that go routine wait until Reader() ends and then calls its own Close() function, so that the two are not tightly coupled.

That's the issue right now. The `Reader` only ends once the context is cancelled. This could be solved by the `EventListenerConnection` having its own context.

Reply:
> I don't understand. Why does `SetWriteDeadline` need a context? It takes a `time.Time`.

I was thinking one could extract the context deadline time and then pass it to the SetWriteDeadline, that is possible, but passing a time.Time works fine too.

> That's the issue right now. The Reader only ends once the context is cancelled. This could be solved by the EventListenerConnection having its own context.

The reader can end if the client disconnects too, so in that case we'd need the caller of the Reader to handle that too, but yes passing a context to Reader for cancellation from the server makes sense too (although I had thought that would be done by calling Close() on the connection, implicitly ending the Reader).